crates/transport/src/layers/retry.rs

```diff
@@ -0,0 +1,296 @@
+                            // Join batch_success and batch_errs
+                            let mut batch_calls = batch_success;
+                            batch_calls.append(&mut batch_errs);
+                            return Ok(ResponsePacket::from(batch_calls));
+                        }
+
+                        if !batch_errs.is_empty() && batch_success.is_empty() {
+                            return Ok(ResponsePacket::from(batch_errs));
+                        }
+
+                        return Err(TransportErrorKind::custom_str(&format!(
+                            "Max retries exceeded {}",
+                            err
+                        )));
+                    }
+                    trace!(%err, "retrying request");
+
+                    let current_queued_reqs = this.requests_enqueued.load(Ordering::SeqCst) as u64;
+
+                    // try to extract the requested backoff from the error or compute the next
+                    // backoff based on retry count
+                    let backoff_hint = this.policy.backoff_hint(&err);
+                    let next_backoff = backoff_hint.unwrap_or_else(|| this.initial_backoff());
+
+                    // requests are usually weighted and can vary from 10 CU to several 100 CU,
+                    // cheaper requests are more common some example alchemy
+                    // weights:
+                    // - `eth_getStorageAt`: 17
+                    // - `eth_getBlockByNumber`: 16
+                    // - `eth_newFilter`: 20
+                    //
+                    // (coming from forking mode) assuming here that storage request will be the
+                    // driver for Rate limits we choose `17` as the average cost
+                    // of any request
+                    const AVG_COST: u64 = 17u64;
+                    let seconds_to_wait_for_compute_budget = compute_unit_offset_in_secs(
+                        AVG_COST,
+                        this.compute_units_per_second,
+                        current_queued_reqs,
+                        ahead_in_queue,
+                    );
+                    let total_backoff = next_backoff
+                        + std::time::Duration::from_secs(seconds_to_wait_for_compute_budget);
+
+                    trace!(
+                        total_backoff_millis = total_backoff.as_millis(),
+                        budget_backoff_millis = seconds_to_wait_for_compute_budget * 1000,
+                        default_backoff_millis = next_backoff.as_millis(),
+                        backoff_hint_millis = backoff_hint.map(|d| d.as_millis()),
+                        "(all in ms) backing off due to rate limit"
+                    );
+
+                    tokio::time::sleep(total_backoff).await;
+                } else {
+                    if timeout_retries < this.max_timeout_retries {
+                        timeout_retries += 1;
+                        continue;
+                    }
```

Comment:
this will always retry errors, such as reverts which are valid responses

we should remove this together with this setting

see also https://github.com/foundry-rs/foundry/pull/8141